_posts/2025-02-04-introduce-bitmap-filtering-feature.md

```diff
@@ -0,0 +1,144 @@
+- `terms` lookup is enhanced to fetch values from stored fields instead of the entire `_source`.
+
+## Example: Filtering a customer’s purchased products
+
+Suppose you run an e-commerce marketplace with 1 million products and 100,000 customers. Each customer has a digital library containing their purchased products. You maintain a bitmap for each customer representing their product ownership. Using bitmap filtering, you can efficiently retrieve the products owned by a specific customer as follows:
+
+```json
+POST products/_search
+{
+  "query": {
+    "terms": {
+      "product_id": {
+        "index": "customers",
+        "id": "customer123",
+        "path": "customer_filter",
+        "store": true
+      },
+      "value_type": "bitmap"
+    }
+  }
+}
+```
+
+In this example, the bitmap filter is applied to the `product_id` field in the `products` index to retrieve products owned by a specific customer. The bitmap filter data is stored in the `customers` index under the document ID `customer123`, in the field `customer_filter`. This binary field is optimized for fast retrieval and efficient processing. During query execution, the bitmap filter is loaded into memory and applied to the filtering operation.
+
+In addition to using a `terms` lookup, as shown in the previous example, you can provide the bitmap directly in the query:
+
+```json
+POST products/_search
+{
+  "query": {
+    "terms": {
+      "product_id": ["<base64-encoded-bitmap>"],
+      "value_type": "bitmap"
+    }
+  }
+}
+```
+
+In this case, the bitmap must be base64-encoded before being included in the query. This approach is useful when you have precomputed bitmaps or need to perform bitmap operations on the client side before querying OpenSearch.
+
+## Key advantages of bitmap filtering
+
+Bitmap filtering integrates seamlessly with OpenSearch’s existing query infrastructure:
+
+- The `value_type: "bitmap"` parameter allows you to specify bitmap filters in `terms` queries.
+- Enhanced `terms` lookup enables efficient retrieval of stored bitmap filters.
+-  You can combine bitmap filters with other query types in Boolean queries, making them a flexible tool for large-scale filtering.
+
+These enhancements enable you to adopt bitmap filtering with minimal changes to your existing OpenSearch queries while gaining significant performance benefits for large-scale filtering operations.
+
+## Performance benchmarks
+
+We conducted performance tests on an index containing 100 million documents, comparing different filtering approaches across filter sizes ranging from 100 to 10 million random IDs.
+
+We compared the following approaches:
+
+- Query using a list of IDs (traditional `terms` query) with document values (OpenSearch 2.17)
+- Query using a list of IDs (traditional `terms` query) with an indexed field (OpenSearch 2.17)
+- Query using bitmap filtering with document values (OpenSearch 2.17)
+- Query using bitmap filtering with an indexed field (OpenSearch 2.19)
```

Comment:
```suggestion
- Query using a list of IDs (standard `terms` query) on a document values only field (OpenSearch 2.17)
- Query using a list of IDs (standard `terms` query) on an indexed field (OpenSearch 2.17)
- Query using bitmap filtering on a document values only field (OpenSearch 2.17)
- Query using bitmap filtering on an indexed field (OpenSearch 2.19)
```

Context: the field here is integer type, by default it's indexed and has doc value enabled, what I did is disable the index, and only keep doc value.

Propose to use standard instead of traditional, as default terms query is not outdated.